### PR TITLE
Confirmaton to save changes in editor

### PR DESCRIPTION
**Pull request by aumayr**
This tries to implement the first part of #605: Adding a confirmation dialog when navigation away from the Editor when there are unsaved changes.

Ref #605

### CODE REVIEW

**Comment by aumayr** [comment]
The current implementation lacks one crucial part: When navigation happens within the page (with the async routing mechanism), this does not get reset in any way. 

@yagebu Is there a way to get an event/notification in JS when the user navigates away from the Editor page on an internal link, so this dialog can be shown/reset accordingly?

**Comment by yagebu** [comment]
> Is there a way to get an event/notification in JS when the user navigates away from the Editor page on an internal link

You can add one to Router.navigate in router.js

**Comment by aumayr** [comment]
OK this is architecturally not simple… in order to keep `Router.navigate` generic there needs to be a mechanism to register callbacks for `beforeNavigate`, so the callback (in this case registered by `editor.js` can cancel the `navigate` if dirty.

File: fava/static/javascript/editor.js [hunk 1]
```diff
@@ -254,6 +255,26 @@ export default function initSourceEditor(name) {
       });
     }
   });
+
+  const leaveMessage = 'There are unsaved changes. Are you sure you want to leave?';
+
+  router.on('navigate', () => {
+    if (!editor.getDoc().isClean()) {
+      const leave = confirm(leaveMessage); // eslint-disable-line no-restricted-globals,no-alert
```
Comment (yagebu) [inline fava/static/javascript/editor.js]: If you use `window.confirm`, one of the warnings should go away

File: fava/static/javascript/editor.js [hunk 2]
```diff
@@ -254,6 +255,26 @@ export default function initSourceEditor(name) {
...
+      if (leave) {
+        editor.getDoc().markClean();
```
Comment (yagebu) [inline fava/static/javascript/editor.js]: This seems wrong

File: fava/static/javascript/editor.js [hunk 3]
```diff
@@ -254,6 +255,26 @@ export default function initSourceEditor(name) {
...
+      }
+      return !leave;
+    }
+    return false;
+  });
+
+  window.onbeforeunload = () => {
```
Comment (yagebu) [inline fava/static/javascript/editor.js]: This and the other event handler persist when one navigates away from the editor. Better use window.addEventListener and add the event handlers once globally (on page-init)

File: fava/static/javascript/editor.js (hunk 2 above)
Comment (aumayr) [reply fava/static/javascript/editor.js]: When navigating away from the (dirty) editor, and then navigating back, the change made before will be gone, but the editor is still marked dirty.

**Comment by aumayr** [comment]
> Also, I'd prefer to just keep the single event handler `e` from events.js.

In 97d4e9f I had to introduce a return value for callbacks to events, because the router (calling `e.trigger(...)` needs to know (from the editor) if the navigation should be allowed.

File: fava/static/javascript/editor.js (hunk 3 above)
Comment (aumayr) [reply fava/static/javascript/editor.js]: The check if the editor is dirty needs a reference to the `editor`-variable, pointing to the CodeMirror instance running the editor. I don't know how to get this reference when putting this code in a global scope without using a global variable (or is that what you suggest?).

Comment (yagebu) [reply fava/static/javascript/editor.js]: This doesn't need a global variable, one within the scope of this file is enough. Right now the above attaches tons of event handlers even though at most one is needed

**Comment by yagebu** [comment]
I've moved the place where we trigger the event to `router.loadURL` so that it also interrupts reloads etc. @aumayr: I think this should be ready, but could you test it to check if there's any edge cases this code might be missing?

**Comment by aumayr** [comment]
Three issues:

- JS-Exception `Uncaught TypeError: Cannot create property 'interrupt' on string 'http://fava:5000/aumayr/statistics/'`
- When navigating away from a "dirty" Editor, the popup shows twice.
- The popup also applies/shows when editing a Journal Entry (clicking on the hash, ...)

**Comment by yagebu** [comment]
I can reproduce none of these issues (in neither Chrome nor Firefox). Can you try to debug them?

**Comment by aumayr** [comment]
Arrr, fault on my part. A `git reset --hard origin/editor-dirty-alert-605` and `make` fixed all the issues 👍
